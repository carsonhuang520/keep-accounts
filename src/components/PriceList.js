import React from 'react'
import PropTypes from 'prop-types'
import Ionicon from 'react-ionicons'

function PriceList(props) {
  const { items, onModifyItem, onDeleteItem } = props
  return (
    <ul className="list-group list-group-flush mt-4">
      {items.map((item) => {
        return (
          <li
            key={item.id}
            className="list-group-item d-flex 
          justify-content-between align-items-center"
          >
            <span className="col-1">
              <Ionicon
                className="rounded-circle"
                fontSize="30px"
                style={{ backgroundColor: '#347eff', padding: '5px' }}
                color={'#fff'}
                icon={item.category.iconName}
              />
            </span>
            <span className="col-5">{item.title}</span>
            <span className="col-2 font-weight-bold">
              {item.category.type === 'outcome' ? '-' : ''}
              {item.price}å…ƒ
            </span>
            <span className="col-2">{item.date}</span>
            <a
              className="col-1"
              role="button"
              href="/"
              onClick={(event) => {
                event.preventDefault()
                onModifyItem(item)
              }}
            >
              <Ionicon
                className="rounded-circle"
                fontSize="30px"
                style={{ backgroundColor: '#28a745', padding: '5px' }}
                color={'#fff'}
                icon="ios-create-outline"
              />
            </a>
            <a
              className="col-1"
              role="button"
              href="/"
              onClick={(event) => {
                event.preventDefault()
                onDeleteItem(item)
              }}
            >
              <Ionicon
                className="rounded-circle"
                fontSize="30px"
                style={{ backgroundColor: '#dc3545', padding: '5px' }}
                color={'#fff'}
                icon="ios-close"
              />
            </a>
          </li>
        )
      })}
    </ul>
  )
}
PriceList.propTypes = {
  items: PropTypes.array.isRequired,
  onModifyItem: PropTypes.func.isRequired,
  onDeleteItem: PropTypes.func.isRequired,
}
export default PriceList
