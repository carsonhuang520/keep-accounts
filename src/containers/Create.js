import React, { Component } from 'react'
import 'bootstrap/dist/css/bootstrap.min.css'
import { Tabs, Tab } from '../components/Tabs'
import CategoryList from '../components/CategoryList'
import PriceForm from '../components/PriceForm'
import withContext from '../WithContext'
import { INCOME, OUTCOME } from '../utility'
const tabsText = [OUTCOME, INCOME]
class Create extends Component {
  constructor(props) {
    super(props)
    this.state = {
      selectedTab: OUTCOME,
      selectedCategory: null,
      validationPassed: true,
    }
  }
  componentDidMount() {
    const { id } = this.props.match.params
    this.props.actions.getEditData(id).then((data) => {
      const { editItem, categories } = data
      this.setState({
        selectedTab: id && editItem ? categories[editItem.cid].type : OUTCOME,
        selectedCategory: id && editItem ? categories[editItem.cid] : null,
      })
    })
  }
  tabChange = (index) => {
    this.setState({
      selectedTab: tabsText[index],
    })
  }
  changeCategory = (category) => {
    this.setState({
      selectedCategory: category,
    })
  }
  handleSubmit = (item, isEdit) => {
    if (!this.state.selectedCategory) {
      this.setState({
        validationPassed: false,
      })
      return
    }
    if (!isEdit) {
      this.props.actions.createItem(item, this.state.selectedCategory.id)
    } else {
      this.props.actions.updateItem(item, this.state.selectedCategory.id)
    }
    this.props.history.push('/')
  }
  cancelSubmit = () => {
    this.props.history.push('/')
  }
  render() {
    const { data } = this.props
    const { id } = this.props.match.params
    const { categories, items } = data
    const { selectedTab, selectedCategory, validationPassed } = this.state
    const editItem = id && items[id] ? items[id] : {}
    const tabIndex = tabsText.findIndex((text) => text === selectedTab)
    const categoriesByType = Object.keys(categories)
      .filter((id) => categories[id].type === selectedTab)
      .map((id) => categories[id])
    return (
      <div>
        <Tabs
          activeIndex={tabIndex}
          onTabChange={(index) => this.tabChange(index)}
        >
          <Tab>支出</Tab>
          <Tab>收入</Tab>
        </Tabs>
        <CategoryList
          categories={categoriesByType}
          selectedCategory={selectedCategory}
          onSelectCategory={this.changeCategory}
        />
        <PriceForm
          item={editItem}
          onFormSubmit={this.handleSubmit}
          onCancelSubmit={this.cancelSubmit}
        />
        {!validationPassed && (
          <div className="alert alert-danger mt-5" role="alert">
            请选择分类信息
          </div>
        )}
      </div>
    )
  }
}

export default withContext(Create)
