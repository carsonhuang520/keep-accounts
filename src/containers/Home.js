import React, { Component, Fragment } from 'react'
import 'bootstrap/dist/css/bootstrap.min.css'
import PriceList from '../components/PriceList'
import ViewTab from '../components/ViewTab'
import TotalPrice from '../components/TotalPrice'
import MonthPicker from '../components/MonthPicker'
import CreateBtn from '../components/CreateBtn'
import { LIST_VIEW, CHART_VIEW, INCOME, OUTCOME } from '../utility'
import withContext from '../WithContext'
import Loader from '../components/Loader'
import PieChart from '../components/PieChart'

const generateChartDataByCategory = (items, type) => {
  let chartData = {}
  items
    .filter((item) => item.category.type === type)
    .forEach((item) => {
      if (chartData[item.cid]) {
        chartData[item.cid].price += item.price * 1
        chartData[item.cid].items.push([item.id])
      } else {
        chartData[item.cid] = {
          name: item.category.name,
          price: item.price * 1,
          items: [item.id],
        }
      }
    })
  const legendData = Object.keys(chartData).map((item) => chartData[item].name)
  const seriesData = Object.keys(chartData).map((item) => ({
    name: chartData[item].name,
    value: chartData[item].price,
  }))
  return {
    legendData,
    seriesData,
  }
}

class Home extends Component {
  constructor(props) {
    super(props)
    this.state = {
      activeTab: LIST_VIEW,
    }
  }
  componentDidMount() {
    this.props.actions.initData()
  }
  changeDate = (year, month) => {
    this.props.actions.selectNewMonth(year, month)
  }
  changeTab = (view) => {
    this.setState({
      activeTab: view,
    })
  }
  addItem = () => {
    this.props.history.push('/create')
  }
  deleteItem = (item) => {
    this.props.actions.deleteItem(item)
  }
  modifyItem = (item) => {
    this.props.history.push(`/update/${item.id}`)
  }
  render() {
    const { data } = this.props
    const { items, categories, currentDate, isLoading } = data
    const { activeTab } = this.state
    const itemsWithCategory = Object.keys(items).map((id) => {
      let item = items[id]
      item.category = categories[item.cid]
      return item
    })
    const outcomeData = generateChartDataByCategory(itemsWithCategory, OUTCOME)
    const incomeData = generateChartDataByCategory(itemsWithCategory, INCOME)
    let outcome = 0
    let income = 0
    itemsWithCategory.forEach((item) => {
      if (item.category.type === INCOME) {
        income += item.price
      } else {
        outcome += item.price
      }
    })
    return (
      <div>
        <div className="row header">
          <div className="col">
            <MonthPicker
              year={currentDate.year}
              month={currentDate.month}
              onChange={(year, month) => this.changeDate(year, month)}
            />
          </div>
          <div className="col">
            <TotalPrice income={income} outcome={outcome} />
          </div>
        </div>
        <div className="content">
          <ViewTab
            activeTab={activeTab}
            onTabChange={(view) => this.changeTab(view)}
          />
          <CreateBtn className="mb-5" onClick={this.addItem} />
          {isLoading ? (
            <Loader />
          ) : (
            <Fragment>
              {activeTab === LIST_VIEW && itemsWithCategory.length > 0 && (
                <PriceList
                  items={itemsWithCategory}
                  onModifyItem={(item) => {
                    this.modifyItem(item)
                  }}
                  onDeleteItem={(item) => {
                    this.deleteItem(item)
                  }}
                />
              )}
              {activeTab === LIST_VIEW && itemsWithCategory.length === 0 && (
                <div className="alert alert-light text-center no-record">
                  您还没有任何记账记录
                </div>
              )}
              {activeTab === CHART_VIEW && itemsWithCategory.length > 0 && (
                <Fragment>
                  {outcomeData.seriesData.length === 0 ? (
                    <h3
                      style={{ textAlign: 'center', fontSize: '24px' }}
                      className="mx-3"
                    >
                      本月没有支出
                    </h3>
                  ) : (
                    <PieChart
                      el="outcome"
                      title="本月支出"
                      name="支出"
                      legendData={outcomeData.legendData}
                      seriesData={outcomeData.seriesData}
                    />
                  )}
                  {incomeData.seriesData.length === 0 ? (
                    <h3
                      style={{ textAlign: 'center', fontSize: '24px' }}
                      className="mx-3"
                    >
                      本月没有收入
                    </h3>
                  ) : (
                    <PieChart
                      el="income"
                      title="本月收入"
                      name="收入"
                      legendData={incomeData.legendData}
                      seriesData={incomeData.seriesData}
                    />
                  )}
                </Fragment>
              )}
              {activeTab === CHART_VIEW && itemsWithCategory.length === 0 && (
                <div className="alert alert-light text-center no-record">
                  您还没有任何记账记录
                </div>
              )}
            </Fragment>
          )}
        </div>
      </div>
    )
  }
}

export default withContext(Home)
