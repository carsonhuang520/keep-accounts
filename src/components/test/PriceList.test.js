import { shallow } from 'enzyme'
import React from 'react'
import Ionicon from 'react-ionicons'
import { items, categories } from '../../containers/Home'
import PriceList from '../PriceList'

const itemsWithCategory = items.map((item) => {
  item.category = categories[item.cid]
  return item
})

const props = {
  items: itemsWithCategory,
  onModifyItem: jest.fn(),
  onDeleteItem: jest.fn(),
}

let wrapper
describe('test PriceList component', () => {
  beforeEach(() => {
    wrapper = shallow(<PriceList {...props} />)
  })
  it('should render the component to match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
  it('should render correct prices length', () => {
    expect(wrapper.find('.list-group-item').length).toEqual(
      itemsWithCategory.length
    )
  })
  it('should render correct icon and price for each item', () => {
    const iconList = wrapper.find('.list-group-item').first().find(Ionicon)
    expect(iconList.length).toEqual(3)
    expect(iconList.first().props().icon).toEqual(
      itemsWithCategory[0].category.iconName
    )
  })
  it('should trigger the correct function callbacks', () => {
    const firstItem = wrapper.find('.list-group-item').first()
    firstItem
      .find('a')
      .first()
      .simulate('click', { preventDefault: () => {} })
    expect(props.onModifyItem).toHaveBeenCalled()
    firstItem
      .find('a')
      .last()
      .simulate('click', { preventDefault: () => {} })
    expect(props.onDeleteItem).toHaveBeenCalled()
  })
})
