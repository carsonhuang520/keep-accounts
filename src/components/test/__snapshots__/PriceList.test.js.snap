// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PriceList component should render the component to match snapshot 1`] = `ShallowWrapper {}`;
